cloud/services/subnets/subnets_test.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subnets
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/golang/mock/gomock"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha2"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/routetables/mock_routetables"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/securitygroups/mock_securitygroups"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/subnets/mock_subnets"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestReconcileSubnets(t *testing.T) {
+	testcases := []struct {
+		name        string
+		subnetSpec  Spec
+		vnetSpec    *infrav1.VnetSpec
+		expectError bool
+		expect      func(m *mock_subnets.MockClientMockRecorder, m1 *mock_routetables.MockClientMockRecorder, m2 *mock_securitygroups.MockClientMockRecorder)
+	}{
+		{
+			name: "subnet does not exist",
+			subnetSpec: Spec{
+				Name:                "my-subnet",
+				CIDR:                "10.0.0.0/16",
+				VnetName:            "my-vnet",
+				RouteTableName:      "my-subent_route_table",
+				SecurityGroupName:   "my-sg",
+				Role:                infrav1.SubnetNode,
+				InternalLBIPAddress: "10.0.0.10",
+			},
+			vnetSpec:    &infrav1.VnetSpec{Name: "my-vnet"},
+			expectError: false,
+			expect: func(m *mock_subnets.MockClientMockRecorder, m1 *mock_routetables.MockClientMockRecorder, m2 *mock_securitygroups.MockClientMockRecorder) {
+				m.Get(context.TODO(), "", "my-vnet", "my-subnet").
+					Return(network.Subnet{}, autorest.NewErrorWithResponse("", "", &http.Response{StatusCode: 404}, "Not found"))
+
+				m1.Get(context.TODO(), "my-rg", "my-subent_route_table").
+					Return(network.RouteTable{}, nil)
+
+				m2.Get(context.TODO(), "my-rg", "my-sg").
+					Return(network.SecurityGroup{}, nil)
+
+				m.CreateOrUpdate(context.TODO(), "", "my-vnet", "my-subnet", gomock.AssignableToTypeOf(network.Subnet{}))
+			},
+		},
+		{
+			name: "vnet was provided but subnet is missing",
```

Comment:
can we also add a test case for "vnet was provided and subnet exists"

Reply:
will do